File: examples/services/node/adapter.js
#!/usr/bin/env node
/**
 * Maestro gRPC Adapter for Next.js API Routes
 * Wraps Next.js API routes to be callable via gRPC
 */

const grpc = require('@grpc/grpc-js');
const protoLoader = require('@grpc/proto-loader');
const axios = require('axios');

class NextJSAdapter {
  constructor(baseUrl = 'http://localhost:3000') {
    this.baseUrl = baseUrl;
  }

  async execute(call, callback) {
    try {
      const { method, payload, correlation_id } = call.request;
      const data = JSON.parse(payload || '{}');
      
      console.log(`Executing ${method} with correlation_id: ${correlation_id}`);
      
      // Map gRPC method to Next.js API route
      const routeMap = {
        'CreateCustomer': '/api/customers',
        'DeleteCustomer': '/api/customers/[id]',
        'SendTemplate': '/api/emails/send',
        'SendOrderConfirmation': '/api/notifications/order',
      };
      
      let endpoint = routeMap[method] || `/api/${method.toLowerCase()}`;
      
      // Replace route parameters
      Object.keys(data).forEach(key => {
        endpoint = endpoint.replace(`[${key}]`, data[key]);
      });
      
      // Determine HTTP method and make request
      let response;
      if (method.startsWith('Create') || method.startsWith('Send')) {
        response = await axios.post(`${this.baseUrl}${endpoint}`, data);
      } else if (method.startsWith('Delete')) {
        response = await axios.delete(`${this.baseUrl}${endpoint}`);
      } else if (method.startsWith('Update')) {
        response = await axios.put(`${this.baseUrl}${endpoint}`, data);
      } else {
        response = await axios.get(`${this.baseUrl}${endpoint}`, { params: data });
      }
      
      callback(null, {
        success: true,
        data: JSON.stringify(response.data),
        error: ''
      });
      
    } catch (error) {
      console.error(`Error executing ${call.request.method}:`, error.message);
      callback(null, {
        success: false,
        data: '{}',
        error: error.message
      });
    }
  }
  
  async compensate(call, callback) {
    // Reuse execute for compensation
    return this.execute(call, callback);
  }
  
  async healthCheck(call, callback) {
    try {
      const response = await axios.get(`${this.baseUrl}/api/health`, { timeout: 5000 });
      callback(null, {
        healthy: response.status === 200,
        message: 'Service is healthy'
      });
    } catch (error) {
      callback(null, {
        healthy: false,
        message: 'Service is not responding'
      });
    }
  }
}

// Start gRPC server
if (require.main === module) {
  const PROTO_PATH = __dirname + '/../../../proto/maestro.proto';
  
  const packageDefinition = protoLoader.loadSync(PROTO_PATH, {
    keepCase: true,
    longs: String,
    enums: String,
    defaults: true,
    oneofs: true
  });
  
  const maestroProto = grpc.loadPackageDefinition(packageDefinition).maestro.v1;
  const server = new grpc.Server();
  const adapter = new NextJSAdapter();
  
  server.addService(maestroProto.MaestroService.service, {
    Execute: adapter.execute.bind(adapter),
    Compensate: adapter.compensate.bind(adapter),
    HealthCheck: adapter.healthCheck.bind(adapter)
  });
  
  const PORT = process.env.GRPC_PORT || '50052';
  server.bindAsync(
    `0.0.0.0:${PORT}`,
    grpc.ServerCredentials.createInsecure(),
    () => {
      console.log(`Next.js Adapter listening on port ${PORT}`);
      server.start();
    }
  );
}

module.exports = NextJSAdapter;

File: examples/services/python/adapter.py
#!/usr/bin/env python3
"""
Maestro gRPC Adapter for FastAPI
Wraps existing FastAPI endpoints to be callable via gRPC
"""

import grpc
import json
import requests
from concurrent import futures
from typing import Any, Dict
import logging

logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

class FastAPIAdapter:
    def __init__(self, base_url: str = "http://localhost:8000"):
        self.base_url = base_url
        
    def Execute(self, request, context):
        """Execute a method on the FastAPI service"""
        try:
            # Parse the request
            method = request.method
            payload = json.loads(request.payload) if request.payload else {}
            
            logger.info(f"Executing {method} with payload: {payload}")
            
            # Map gRPC method to FastAPI endpoint
            endpoint_map = {
                "CreateUser": "/api/users",
                "DeleteUser": "/api/users/{user_id}",
                "GetUser": "/api/users/{user_id}",
                "UpdateUser": "/api/users/{user_id}",
            }
            
            endpoint = endpoint_map.get(method, f"/api/{method.lower()}")
            
            # Replace path parameters
            for key, value in payload.items():
                endpoint = endpoint.replace(f"{{{key}}}", str(value))
            
            # Determine HTTP method
            if method.startswith("Create"):
                response = requests.post(f"{self.base_url}{endpoint}", json=payload)
            elif method.startswith("Delete"):
                response = requests.delete(f"{self.base_url}{endpoint}")
            elif method.startswith("Update"):
                response = requests.put(f"{self.base_url}{endpoint}", json=payload)
            else:
                response = requests.get(f"{self.base_url}{endpoint}", params=payload)
            
            # Return response
            return {
                "success": response.status_code < 400,
                "data": response.json() if response.text else {},
                "error": str(response.text) if response.status_code >= 400 else ""
            }
            
        except Exception as e:
            logger.error(f"Error executing {request.method}: {e}")
            return {
                "success": False,
                "data": {},
                "error": str(e)
            }
    
    def Compensate(self, request, context):
        """Execute compensation logic"""
        return self.Execute(request, context)
    
    def HealthCheck(self, request, context):
        """Check service health"""
        try:
            response = requests.get(f"{self.base_url}/health", timeout=5)
            return {
                "healthy": response.status_code == 200,
                "message": "Service is healthy"
            }
        except:
            return {
                "healthy": False,
                "message": "Service is not responding"
            }

if __name__ == "__main__":
    # Start gRPC server
    server = grpc.server(futures.ThreadPoolExecutor(max_workers=10))
    adapter = FastAPIAdapter()
    
    # Register the service (you'll need to generate Python proto files)
    # maestro_pb2_grpc.add_MaestroServiceServicer_to_server(adapter, server)
    
    server.add_insecure_port('[::]:50051')
    server.start()
    logger.info("FastAPI Adapter listening on port 50051")
    server.wait_for_termination()